File: lib/devcenter/commands/pull.rb
module Devcenter::Commands

  class Pull < Base

    def initialize(*args)
      @slug = slug_from_article_url(args[0])
      @force_overwrite = args[1]
      super
    end

    def validate
      unless @slug && !@slug.strip.empty?
        @validation_errors << 'Please provide an article slug or full (e.g: ps or https://devcenter.heroku.com/articles/ps)'
      end
    end

    def run
      response = Devcenter::Client.get(path: article_api_path(@slug))
      article_received = response.ok? && response.body['article'] && response.body['article']['id']
      article_not_found!(@slug) unless article_received

      article = response.body['article']
      metadata = {'title' => article['title'], 'id' => article['id']}
      file_path = md_file_path(@slug)

      unless @force_overwrite
        cancel_save = File.exists?(file_path) && !agree("The file #{file_path} already exists - overwrite? (yes/no)")
        return if cancel_save
      end

      file_content = [Psych.dump(metadata), article['content']].join("\n\n")
      write_file(file_path, file_content)
      say "\"#{metadata['title']}\" article saved as #{file_path}"
    end

  end
end

File: src/commands/devcenter/pull.ts
import {Command} from '@heroku-cli/command'
import {confirm} from '@heroku/heroku-cli-util/hux'
import {Args, Flags} from '@oclif/core'
import createDebug from 'debug'
import {existsSync, writeFileSync} from 'node:fs'
import {stringify as stringifyYaml} from 'yaml'

import {formatArticleNotFoundMessage} from '../../lib/article-not-found.js'
import {fetchArticleJsonForSlug} from '../../lib/article-resolve.js'
import {DevcenterClient} from '../../lib/devcenter-client.js'
import {
  articleApiPath,
  getDevcenterBaseUrl,
  mdFilePath,
  slugFromArticleUrl,
} from '../../lib/paths.js'

const dbg = createDebug('devcenter:pull')

export default class Pull extends Command {
  static args = {
    slugOrUrl: Args.string({
      description: 'article slug or full Dev Center article URL',
      required: true,
    }),
  }
  static description = 'save a local copy of a Dev Center article'
  static flags = {
    force: Flags.boolean({
      char: 'f',
      description: 'overwrite an existing local file without prompting',
    }),
  }

  async run(): Promise<void> {
    const {args, flags} = await this.parse(Pull)
    const raw = args.slugOrUrl.trim()
    const slug = slugFromArticleUrl(raw).trim()
    const client = new DevcenterClient()
    dbg(`baseUrl=${getDevcenterBaseUrl()} path=${articleApiPath(slug)} expectedSlug=${slug}`)

    const token = await this.heroku.getAuth()
    const article = await fetchArticleJsonForSlug(client, slug, token, dbg)
    if (!article) {
      const msg = await formatArticleNotFoundMessage(client, slug)
      this.error(msg, {exit: 1})
    }

    const metadata = {id: article.id, title: article.title}
    const filePath = mdFilePath(slug)

    if (!flags.force && existsSync(filePath)) {
      const shouldOverwrite = await confirm(`The file ${filePath} already exists - overwrite?`)
      if (!shouldOverwrite) {
        return
      }
    }

    const yamlBlock = stringifyYaml(metadata).trimEnd()
    const fileContent = `${yamlBlock}\n\n${article.content}`
    writeFileSync(filePath, fileContent, 'utf8')
    this.log(`"${metadata.title}" article saved as ${filePath}`)
  }
}
